File: Wayne/WayneCompiler/src/Dfir_representation/IConstant.java
package Dfir_representation;

import java.util.Vector;


public  class IConstant extends JNode{
	   private String value;
	   private IDataType iDataType;
	   public IConstant(int nodeId, int parentId, Vector<ITerminal> inputTerminals,Vector<ITerminal> outputTerminals, String value, IDataType iDataType){
		   super(nodeId,parentId,inputTerminals,outputTerminals);
		   this.value = value;
		   this.iDataType = iDataType;
		   JNodeType = "IConstant";
	   }
	   public String getValue(){
		   if(this.getIDataType().getBase() == BaseDataType.valueOf("IArray") || this.getIDataType().getBase() == BaseDataType.valueOf("IVariableSizedArray")  ||this.getIDataType().getBase() == BaseDataType.valueOf("IFixedSizeArray") ){
				String s = this.value;
				return s.replace('[', '{').replace(']', '}');
			}else if(this.getIDataType().getBase() == BaseDataType.valueOf("IBoolean")){
				if(this.value.equals("False")){
					return "false";
				}else if(this.value.equals("True")){
					return "true";
				}
				
			}
		   return this.value;
	   }
	   
	   public String getReal(){
			if(this.iDataType.getBase() == BaseDataType.valueOf("IComplex")){
				return this.value.split("\\+")[0].trim();
			}
			return "IComplex Real Error!!!";
		}
	   public String getImag(){
			if(this.iDataType.getBase() == BaseDataType.valueOf("IComplex")){
				return this.value.split("\\+")[1].split("i")[0].trim();
			}
			return "IComplex Real Error!!!";
		}
	   
	   public IDataType getIDataType(){
		   return this.iDataType;
	   }
	   public void print(){
		   System.out.println("IConstant begin");
		   System.out.println("NodeId: " + this.nodeId + ' ' + "ParentId: " + this.parentId + "\nValue: " + this.value);
		   iDataType.print();
		   for(int i=0; i<outputTerminals.size(); i++){
			   outputTerminals.get(i).print();
		   }
		   System.out.println("IConstant end\n");
	   }
	   public void AddOutputTerminals(ITerminal iTerminal){
		   this.outputTerminals.addElement(iTerminal);
	   }
}

File: Jasmine/JasmineCompiler/Src/DFIR_Compiler/DFIR_Compiler/IConstant.cs
using System;
using System.Collections.Generic;

namespace DFIR_Compiler
{
    public class IConstant : INode
    {
        private string value;
        private IDataType iconstantIDataType;
        public IConstant(int nodeId, int parentId, List<ITerminal> inputTerminals, List<ITerminal> outputTerminals, string Value, IDataType iDataType) : base(nodeId, parentId, inputTerminals, outputTerminals)
        {
            value = Value;
            iconstantIDataType = iDataType;
            INodeType = "IConstant";
        }


        public virtual string GetValue()
        {
            if (GetIDataType().GetBase() == BaseDataType.IBoolean)
            {
                if (value == "False")
                {
                    return "false";
                }
                else if (value == "True")
                {
                    return "true";
                }
            }
            else if (GetIDataType().GetBase() == BaseDataType.IArray || GetIDataType().GetBase() == BaseDataType.IVariableSizedArray || GetIDataType().GetBase() == BaseDataType.IFixedSizeArray)
            {
                string s = value;
                return s.Replace('[', '{').Replace(']', '}');
            }
            return value;
        }

        //Get the real part of the IComplex
        public virtual string GetReal()
        {
            if (iconstantIDataType.GetBase() == BaseDataType.IComplex)
            {
                return value.Split('+')[0].Trim();
            }
            return " Error: IComplex Real!";
        }
        //Get the imaginary part of the IComplex
        public virtual string GetImag()
        {
            if (iconstantIDataType.GetBase() == BaseDataType.IComplex)
            {
                return value.Split('+')[1].Split('i')[0].Trim();
            }
            return " Error: IComplex Imag!";
        }

        public virtual IDataType GetIDataType()
        {
            return iconstantIDataType;
        }

        public override void Print()
        {
            Console.WriteLine("------------------------------------");
            Console.WriteLine("IConstant:");
            Console.WriteLine("NodeId: " + nodeId + ' ' + "ParentId: " + parentId + "\nValue: " + value);
            iconstantIDataType.Print();
            for (int i = 0; i < outputTerminals.Count; i++)
            {
                outputTerminals[i].Print();
            }
        }



    }
}
